lib/auth/winwebauthn/webauthn_windows.go

```diff
@@ -0,0 +1,945 @@
+	}
+
+	// Note that we need to copy bytes out of out if we want to free object.
+	// That's why mkBytesSliceOutOfCbytes is used.
+	defer freeCredentialAttestation(out)
+
+	credential := mkBytesSliceOutOfCbytes(out.CbCredentialId, out.PbCredentialId)
+
+	return &wanlib.CredentialCreationResponse{
+		PublicKeyCredential: wanlib.PublicKeyCredential{
+			Credential: wanlib.Credential{
+				ID:   base64.RawURLEncoding.EncodeToString(credential),
+				Type: string(protocol.PublicKeyCredentialType),
+			},
+			RawID: credential,
+		},
+		AttestationResponse: wanlib.AuthenticatorAttestationResponse{
+			AuthenticatorResponse: wanlib.AuthenticatorResponse{
+				ClientDataJSON: jsonEncodedCd,
+			},
+			AttestationObject: mkBytesSliceOutOfCbytes(out.CbAttestationObject, out.PbAttestationObject),
+		},
+	}, nil
+}
+
+func freeCredentialAttestation(in *_WEBAUTHN_CREDENTIAL_ATTESTATION) error {
+	_, _, err := procWebAuthNFreeCredentialAttestation.Call(
+		uintptr(unsafe.Pointer(in)),
+	)
+	if err != syscall.Errno(0) {
+		return err
+	}
+	return nil
+}
+
+func freeAssertion(in *_WEBAUTHN_ASSERTION) error {
+	_, _, err := procWebAuthNFreeAssertion.Call(
+		uintptr(unsafe.Pointer(in)),
+	)
+	if err != syscall.Errno(0) {
+		return err
+	}
+	return nil
+}
+
+type hresult int32
+
+// checkIfDLLExistsAndGetAPIVersionNumber checks if dll exists and tries to load
+// it's version via API call. This function makes sure to not panic if dll is
+// missing.
+func checkIfDLLExistsAndGetAPIVersionNumber() (int, error) {
+	if err := modWebAuthn.Load(); err != nil {
+		return 0, err
+	}
+	if err := procWebAuthNGetApiVersionNumber.Find(); err != nil {
+		return 0, err
+	}
+	ret, _, err := procWebAuthNGetApiVersionNumber.Call()
+	if err != syscall.Errno(0) {
+		return 0, err
+	}
+	return int(ret), nil
+}
+
+func getErrorName(in hresult, originCode uintptr) error {
+	ret, _, err := procWebAuthNGetErrorName.Call(
+		uintptr(in),
+	)
+	if err != syscall.Errno(0) {
+		return fmt.Errorf("could not check error name for %x because of: %x", ret, err)
+	}
+	errString := windows.UTF16PtrToString((*uint16)(unsafe.Pointer(ret)))
+	return fmt.Errorf("Webauthn err for code %v: %s", originCode, errString)
+}
+
+func isUVPlatformAuthenticatorAvailable() (bool, error) {
+	var out uint32
+	ret, _, err := procWebAuthNIsUserVerifyingPlatformAuthenticatorAvailable.Call(
+		uintptr(unsafe.Pointer(&out)),
+	)
+	if err != syscall.Errno(0) {
+		return false, err
+	}
+	if ret != 0 {
+		return false, getErrorName(hresult(ret), ret)
+	}
+	return out == 1, nil
+}
+
+func (c client) assertOptionsToCType(in protocol.PublicKeyCredentialRequestOptions, loginOpts *LoginOpts) (*_WEBAUTHN_AUTHENTICATOR_GET_ASSERTION_OPTIONS, error) {
+	allowCredList, err := credentialsExToCType(in.AllowedCredentials)
+	if err != nil {
+		return nil, err
+	}
+	var dwVersion uint32
+	switch c.version {
+	case apiVersion1, apiVersion2:
+		dwVersion = 4
+	case apiVersion3:
+		dwVersion = 5
+	case apiVersion4:
+		dwVersion = 6
```

Comment:
A Q about Windows APIs: could we simply set the highest version number we support and let older APIs do the best they can? Or does it not work like that?

Reply:
I am not sure if I understand. 
I think the problem is that there are multiple kind of "versions" used.

First type is API version (v1-v4).
Other type is version of structure in endpoints. (They don't match API version at all).
https://github.com/microsoft/webauthn/blob/7ab979cc833bfab9a682ed51761309db57f56c8c/webauthn.h#L36-L97

Reply:
Ack. Remind me to test as well, my API version is likely older than yours. Meanwhile, any chance you could find some documentation that explains how Windows deals with this?

I think this get considerably easier if we can just set the highest version number we support and move on.

Reply:
ok i was able to test it on older API and you were right, we can set highest number on older API and it still works, I have changed code to only contain newest version

Reply:
Nice.